### PR TITLE
Switch from PyCrypto to Cryptography

### LINKED ISSUE #243: Replace PyCrypto by cryptography
Hi all,

as you may know, _scapy_ uses _PyCrypto_ for some (optional but useful) features.

As stated by _PyCrypto_'s maintainer, [this project is _almost_ dead](https://lists.dlitz.net/pipermail/pycrypto/2014q1/000782.html). It has [not received any activity since june 2014](https://github.com/dlitz/pycrypto/commits/master).

On 2014-02-06 09:51:39, Dwayne Litzenberger wrote:

> I'm thinking of merging with the folks at https://cryptography.io/,  which is covered by an Apache 2.0 license.

Some python libs have started to move from _PyCrypto_ to _cryptography_: https://github.com/paramiko/paramiko/pull/394

Indeed, _cryptography_ is much more active. It has detailed documentation and is based on _openssl_. I also supports pypy, and installation on windows is easier. https://cryptography.io/en/latest/ https://pypi.python.org/pypi/cryptography

It would be a good thing to use _cryptography_ instead of PyCrypto. This concerns the IPsec and TLS layers, mostly.

What do you think?

### PR DESCRIPTION
**Pull request by p-l-**
This is based on #263, with conflicts fixed and CCM support (temporarily I hope) dropped.

What do you think @ThomasFaivre @robin-jarry @guedou and others?

Fixes #263.
Fixes #243.

### CODE REVIEW

**Comment by robin-jarry** [comment]
Hey @p-l- that's wonderful news !

TLS layer seems broken with pypy,  other than that : gogogogo :-)

Acked-by: Robin Jarry <robin.jarry@6wind.com>

**Comment by p-l-** [comment]
@robin-jarry thanks! I'll try to have Pypy tests pass, or I will comment the Pypy tests if I cannot.

**Comment by p-l-** [comment]
I have to disable tests using AES-CCM: see pyca/cryptography#2968 (comments) and pyca/cryptography#1141 (older).

@mtury what do you think?

**Comment by mtury** [comment]
Although it messes up with the TLS support, I think this switch is the right thing to do. I guess I'll keep one legacy branch on my fork using `pycrypto` for CCM use, while I'll update the main TLS branch used for the PR so that it uses `cryptography`.

File: .travis/install.sh [hunk 1]
```diff
@@ -7,17 +7,30 @@ then
...
 
-# Pycrypto 2.7a1 isn't available on PyPi
-if [ "$TEST_COMBINED_MODES" = "yes" ]
-then
-  curl -sL https://github.com/dlitz/pycrypto/archive/v2.7a1.tar.gz | tar xz
-  cd pycrypto-2.7a1
-  python setup.py build
-  $SCAPY_SUDO python setup.py install
+if python --version 2>&1 | grep -q PyPy; then
+  wget https://pypi.python.org/packages/82/f7/d6dfd7595910a20a563a83a762bf79a253c4df71759c3b228accb3d7e5e4/cryptography-1.7.1.tar.gz
```
Comment (guedou) [inline .travis/install.sh]: Is it  possible to use `pip install`" instead of wget/tar/cd ?

File: .travis/install.sh [hunk 2]
```diff
@@ -7,17 +7,30 @@ then
...
+  tar zxf cryptography-1.7.1.tar.gz
+  cd cryptography-1.7.1
+  patch << EOF
+--- setup.py
```
Comment (guedou) [inline .travis/install.sh]: Could you add a comment explaining why this patch is required ?

File: scapy/layers/dot11.py [hunk 3]
```diff
@@ -339,8 +343,12 @@ def post_build(self, p, pay):
                     icv = ""
                 else:
                     icv = p[4:8]
-                c = ARC4.new(self.iv+key)
-                p = p[:4]+c.encrypt(pay)+icv
+                e = Cipher(
```
Comment (guedou) [inline scapy/layers/dot11.py]: I wonder if we should _protect_ the "WEP tests" UT in `regression.uts` in order to make the unit tests run smoothly is cryptography is not installed. What do you think ?

**Comment by guedou** [comment]
I made some cosmetic comments. Otherwise, this PR is good to be merged !

**Comment by p-l-** [comment]
@guedou in the end, I've prevented "crypto" tests from running with Pypy. What do you think?

I had to fix `UTScapy` for that ;-).

**Comment by gpotter2** [comment]
@p-l- I might be wrong, but from what i've seen [here](https://github.com/travis-ci/travis-ci/issues/4756)
it appears that replacing `pypy` by `pypy-5.4.1` in `.travis.yml` fixes it...

**Comment by guedou** [comment]
@p-l @gpotter2 we could try that. The default pypy version is old.

**Comment by p-l-** [comment]
@gpotter2 crypto is tested with "regular" CPython tests anyway, so I don't really mind. Plus it **seems** that so far it only works with `sudo` enabled.

When the regular Pypy in Travis CI is recent enough, we'll enable those tests again.

**Comment by gpotter2** [comment]
Well ok.

BTW, it says that only `pypy` doesn't return the same version with `sudo: false`, but `pypy-5.4.1` will return a newer version anyway.

**Comment by p-l-** [comment]
@guedou conflict fixed!

**Comment by p-l-** [comment]
Thanks for this work @ThomasFaivre, sorry it took us so long to fix the tests, clean-up and merge it.

**Comment by ThomasFaivre** [comment]
No problem! And sorry for CCM and TLS issues :/

**Comment by robin-jarry** [comment]
Yay! 👍

**Pull request merged**